File: src/bowlinggame/BowlingGame.java
package bowlinggame;

import java.util.ArrayList;
import java.util.List;

public class BowlingGame {

    private enum FrameType {

        STRIKE, SPARE, DEFAULT
    };
    private List<Integer> ballScores = new ArrayList<Integer>();

    public int getScore() {
        int score = 0;
        int frame = 0;
        for (int i = 0; i < ballScores.size() && frame < 10; i++) {
            switch (getFrameType(i)) {
                case STRIKE:
                    score += 10 + getBallScore(i + 1);
                    score += getBallScore(i + 2);
                    break;
                case SPARE:
                    score += 10 + getBallScore(i + 2);
                    i++;
                    break;
                default:
                    score += getBallScore(i) + getBallScore(i + 1);
                    i++;
            }
            frame++;
        }
        return score;
    }

    private FrameType getFrameType(int index) {
        FrameType type;

        if (getBallScore(index) == 10) {
            type = FrameType.STRIKE;
        } else if (getBallScore(index) + getBallScore(index + 1) == 10) {
            type = FrameType.SPARE;
        } else {
            type = FrameType.DEFAULT;
        }
        return type;
    }

    private int getBallScore(int index) {
        return (index < ballScores.size()) ? ballScores.get(index) : 0;
    }

    public void pinsDown(int noOfPinsDown) {
        if (noOfPinsDown > 10) {
            throw new IllegalStateException("Bowl score was out of bounds");
        }
        ballScores.add(noOfPinsDown);
    }
}

File: src/main/scala/bowlinggame/BowlingGame.scala
package bowlinggame

import scala.annotation.tailrec

class BowlingGame {
  private var ballScores = List.empty[Int]

  def getScore: Int = {
    @tailrec
    def getScoreRec(score: Int, framesLeft: Int, balls: List[Int]): Int = framesLeft match {
      case 0 => score
      case _ => balls match {
        case 10 :: xs => // Strike
          xs match {
            case b1 :: b2 :: xs2 => getScoreRec(score + 10 + b1 + b2, framesLeft - 1, xs)
            case b1 :: xs2 => getScoreRec(score + 10 + b1, framesLeft - 1, xs)
            case _  => getScoreRec(score + 10, framesLeft - 1, xs)
          }
        case b1 :: b2 :: xs if b1 + b2 == 10 => // Spare
          getScoreRec(score + 10 + xs.headOption.getOrElse(0), framesLeft - 1, xs)
        case b1 :: b2 :: xs => // regular bowl
          getScoreRec(score + b1 + b2, framesLeft - 1, xs)
        case b1 :: Nil => // regular, incomplete bowl
          score + b1
        case _ => score
      }
    }
    getScoreRec(0, 10, ballScores)
  }

  def newRoll(pinsDown: Int): Unit = {
    assert(pinsDown <= 10 && pinsDown >= 0, "bowl must be >=0 and <=10")
    ballScores = ballScores ::: List(pinsDown)
  }
}
